src/OrbitGl/ThreadTrack.cpp

```diff
@@ -405,3 +397,100 @@ float ThreadTrack::GetHeaderHeight() const {
+  float world_timer_x =
+      static_cast<float>(draw_data.world_start_x + normalized_start * draw_data.world_width);
+  return internal::Rect(world_timer_x, world_pos_y, world_timer_width, world_size_y);
+}
+
+[[nodiscard]] static inline uint64_t GetNextPixelBoundaryTimeNs(
+    float world_x, const internal::DrawData& draw_data) {
+  float normalized_x = (world_x - draw_data.world_start_x) / draw_data.world_width;
+  int pixel_x = static_cast<int>(ceil(normalized_x * draw_data.canvas->GetWidth()));
+  return draw_data.min_tick + pixel_x * draw_data.ns_per_pixel;
+}
+
+// We minimize overdraw when drawing lines for small events by discarding events that would just
+// draw over an already drawn pixel line. When zoomed in enough that all events are drawn as boxes,
+// this has no effect. When zoomed  out, many events will be discarded quickly.
+void ThreadTrack::UpdatePrimitives(Batcher* batcher, uint64_t min_tick, uint64_t max_tick,
+                                   PickingMode picking_mode, float z_offset) {
+  UpdatePrimitivesOfSubtracks(batcher, min_tick, max_tick, picking_mode, z_offset);
+  UpdateBoxHeight();
+  visible_timer_count_ = 0;
+  internal::DrawData draw_data = GetDrawData(
+      min_tick, max_tick, z_offset, batcher, time_graph_, collapse_toggle_->IsCollapsed(),
+      app_->selected_text_box(), app_->GetFunctionIdToHighlight());
+
+  absl::MutexLock lock(&scope_tree_mutex_);
+
+  for (auto& [depth, ordered_nodes] : scope_tree_.GetOrderedNodesByDepth()) {
+    auto first_node_to_draw = ordered_nodes.lower_bound(min_tick);
+    if (first_node_to_draw != ordered_nodes.begin()) --first_node_to_draw;
+
+    UpdateDepth(depth);
+    float world_timer_y = GetYFromDepth(depth - 1);
+    uint64_t next_pixel_start_time_ns = min_tick;
+
+    for (auto it = first_node_to_draw; it != ordered_nodes.end() && it->first < max_tick; ++it) {
+      TextBox& text_box = *it->second->GetScope();
+      const TimerInfo& timer_info = text_box.GetTimerInfo();
+      if (timer_info.end() <= next_pixel_start_time_ns) continue;
+      ++visible_timer_count_;
+
+      Color color = GetTimerColor(text_box, draw_data);
+      auto rect = GetTimerInfoRect(timer_info, draw_data, time_graph_, world_timer_y, box_height_);
+      text_box.SetPosAndSize(rect.pos, rect.size);
+      auto user_data = std::make_unique<PickingUserData>(
+          &text_box, [&, batcher](PickingId id) { return this->GetBoxTooltip(*batcher, id); });
+
+      if ((timer_info.end() - timer_info.start()) > draw_data.ns_per_pixel) {
+        SetTimesliceText(timer_info, draw_data.world_start_x, z_offset, &text_box);
+        batcher->AddShadedBox(rect.pos, rect.size, draw_data.z, color, std::move(user_data));
+      } else {
+        batcher->AddVerticalLine(rect.pos, rect.size[1], draw_data.z, color, std::move(user_data));
+      }
+
+      // Use the time at boundary of the next pixel as a threshold to avoid overdraw.
```

Comment:
Using only the next pixel at the boundary and discarding everything earlier looks easier and more elegant. The only issue I can see is that this requires that timers are sorted. 

Otherwise with something like the following draw order (assume that "1" is drawn first), "3" and "2" will not get drawn at all.
3  2  1
 |   |   | 

If we assume that timers are sorted by end time and don't overlap, then they are also, at the same depth, sorted by start time, which will avoid this problem. For GPU debug label timers, there's potentially a problem here, as we know they can overlap, but if this is only in `ThreadTrack`, it should work fine. In either case, it will likely be unnoticeable in practice but that argument is hand-wavy at best.

Of course, it would be good to verify there's no performance regression, but sounds like you are planning to do that anyway.

Reply:
Since we are using ScopeTree here, the timers are guaranteed to be sorted by start time. In the case of ThreadTrack, there is no overlap. But I agree that we need to keep that in mind for other tracks allowing overlap or not using the ScopeTree at all.